In [4]:
import pandas as pd
import numpy as np
import category_encoders as ce
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split


In [5]:
dataset = pd.read_csv("./datasets/datathon_train.csv")

In [22]:
print(dataset.head())

   Id  MONTH  DAY_OF_MONTH  DAY_OF_WEEK             DEPARTING_AIRPORT  \
0   0      1            17            4  Raleigh-Durham International   
1   1      2            27            3        San Jose International   
2   2      8            14            3     Los Angeles International   
3   3      3            16            6  Newark Liberty International   
4   4      6            20            4             Douglas Municipal   

     ORIGIN_CITY_NAME DEST DEST_CITY_NAME DEP_TIME_BLK  DISTANCE  ...  \
0  Raleigh/Durham, NC  JFK   New York, NY    1800-1859       427  ...   
1        San Jose, CA  BOS     Boston, MA    2200-2259      2689  ...   
2     Los Angeles, CA  ATL    Atlanta, GA    0600-0659      1947  ...   
3          Newark, NJ  DTW    Detroit, MI    1400-1459       488  ...   
4       Charlotte, NC  JFK   New York, NY    1500-1559       541  ...   

   PLANE_AGE                   PREVIOUS_AIRPORT  PRCP  SNOW SNWD  TMAX   AWND  \
0          5  Ronald Reagan Washington Na

In [6]:
def get_time(string):
  return int(string[:2])
dataset['DEP_TIME'] = dataset['DEP_TIME_BLK'].apply(get_time)

In [7]:
dataset_clean = dataset.drop(['Id', 'ORIGIN_CITY_NAME', 'DEST_CITY_NAME', 'DEP_TIME_BLK'], axis = 1)

In [8]:
#use binary encoder
cols = ['PREVIOUS_AIRPORT', 'DEST', 'CARRIER_NAME', 'DEPARTING_AIRPORT']
encoder = ce.BinaryEncoder(cols=cols)
dataset_encoded = encoder.fit_transform(dataset_clean)
dataset_encoded = dataset_encoded.dropna()

/home/sae/ENTER/envs/newconda/lib/python3.9/site-packages/category_encoders/utils.py:28: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead
  elif pd.api.types.is_categorical_dtype(cols):
/home/sae/ENTER/envs/newconda/lib/python3.9/site-packages/category_encoders/utils.py:50: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead
  return pd.api.types.is_categorical_dtype(dtype)
/home/sae/ENTER/envs/newconda/lib/python3.9/site-packages/category_encoders/utils.py:50: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead
  return pd.api.types.is_categorical_dtype(dtype)
/home/sae/ENTER/envs/newconda/lib/python3.9/site-packages/category_encoders/utils.py:50: FutureWarning: is_categorical_dtype is deprecated and will be removed in a futu

In [9]:
dataset_encoded

MONTH  DAY_OF_MONTH  DAY_OF_WEEK  DEPARTING_AIRPORT_0  \
0           1            17            4                    0   
1           2            27            3                    0   
2           8            14            3                    0   
3           3            16            6                    0   
4           6            20            4                    0   
...       ...           ...          ...                  ...   
697219      4             4            4                    0   
697220     10            22            2                    0   
697221     10            30            3                    0   
697222     11            27            3                    0   
697223      8            12            1                    0   

        DEPARTING_AIRPORT_1  DEPARTING_AIRPORT_2  DEPARTING_AIRPORT_3  \
0                         0                    0                    0   
1                         0                    0                    0   
2                         0                    0                    0   
3                         0                    0                    0   
4                         0                    0                    0   
...                     ...                  ...                  ...   
697219                    1                    0                    0   
697220                    1                    0                    0   
697221                    0                    0                    0   
697222                    0                    1                    0   
697223                    0                    1                    0   

        DEPARTING_AIRPORT_4  DEPARTING_AIRPORT_5  DEPARTING_AIRPORT_6  ...  \
0                         0                    0                    1  ...   
1                         0                    1                    0  ...   
2                         0                    1                    1  ...   
3                         1                    0                    0  ...   
4                         1                    0                    1  ...   
...                     ...                  ...                  ...  ...   
697219                    0                    1                    1  ...   
697220                    1                    1                    0  ...   
697221                    1                    0                    0  ...   
697222                    1                    0                    1  ...   
697223                    0                    1                    1  ...   

        PREVIOUS_AIRPORT_7  PREVIOUS_AIRPORT_8  PRCP  SNOW  SNWD   TMAX  \
0                        0                   1  0.01   0.0   0.0   49.0   
1                        1                   0  0.21   0.0   0.0   64.0   
2                        1                   1  0.00   0.0   0.0   74.0   
3                        0                   0  0.00   0.0   0.0   56.0   
4                        0                   1  0.19   0.0   0.0   89.0   
...                    ...                 ...   ...   ...   ...    ...   
697219                   1                   1  0.00   0.0   0.0   86.0   
697220                   0                   0  0.00   0.0   0.0   76.0   
697221                   0                   0  0.15   0.0   0.0   65.0   
697222                   1                   1  0.00   0.0   0.0   57.0   
697223                   1                   1  0.00   0.0   0.0  102.0   

         AWND  DEP_DELAY_NEW  IS_DELAYED  DEP_TIME  
0        4.70            2.0           1        18  
1       12.75            0.0           0        22  
2        8.50            0.0           0         6  
3       17.67            0.0           0        14  
4        9.17           79.0           1        15  
...       ...            ...         ...       ...  
697219   8.50            0.0           0         7  
697220   2.91            7.0           1        13  
697221   5.59            0.0       

In [10]:
X_df = dataset_encoded.drop(['DEP_DELAY_NEW', 'IS_DELAYED'], axis = 1)
X = X_df.to_numpy()
y = dataset_encoded['IS_DELAYED'].to_numpy()

In [12]:
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.2, random_state=42)

In [18]:
X_pca = PCA().fit_transform(X_train)
NUM_SELECTED_COLUMNS = 7
selected_columns_train = X_pca[:,:NUM_SELECTED_COLUMNS]

In [16]:
X_pca = PCA().fit_transform(X_test)
NUM_SELECTED_COLUMNS = 7
selected_columns_test = X_pca[:,:NUM_SELECTED_COLUMNS]

In [19]:
#Create and train a decision tree classifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.tree import DecisionTreeClassifier

clf = DecisionTreeClassifier(random_state=42)  # You can customize hyperparameters if needed
clf.fit(selected_columns_train, y_train)

# Step 4: Evaluate the classifier's performance on the testing data
y_pred = clf.predict(selected_columns_test)
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("Classification Report:\n", report)

Accuracy: 0.5222345727706265
Classification Report:
               precision    recall  f1-score   support

           0       0.62      0.54      0.58     83870
           1       0.42      0.49      0.45     55575

    accuracy                           0.52    139445
   macro avg       0.52      0.52      0.51    139445
weighted avg       0.54      0.52      0.53    139445



Accuracy: 0.5898764242194761
Classification Report:
               precision    recall  f1-score   support

           0       0.66      0.65      0.66    105073
           1       0.48      0.50      0.49     69233

    accuracy                           0.59    174306
   macro avg       0.57      0.57      0.57    174306
weighted avg       0.59      0.59      0.59    174306


In [20]:
from sklearn.model_selection import train_test_split, GridSearchCV
param_grid = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

grid_search = GridSearchCV(clf, param_grid, cv=5, scoring='accuracy', n_jobs=-1)
grid_search.fit(X_train, y_train)


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=42), n_jobs=-1,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [None, 10, 20, 30],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10]},
             scoring='accuracy')

In [21]:
best_clf = grid_search.best_estimator_

y_pred = best_clf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print("Best Parameters:", grid_search.best_params_)
print("Accuracy:", accuracy)
print("Classification Report:\n", report)

Best Parameters: {'criterion': 'entropy', 'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 2}
Accuracy: 0.654580659041199
Classification Report:
               precision    recall  f1-score   support

           0       0.67      0.83      0.74     83870
           1       0.61      0.38      0.47     55575

    accuracy                           0.65    139445
   macro avg       0.64      0.61      0.61    139445
weighted avg       0.64      0.65      0.63    139445



Without PCA
Best Parameters: {'criterion': 'entropy', 'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 10}
Accuracy: 0.6529838330292703
Classification Report:
               precision    recall  f1-score   support

           0       0.67      0.84      0.74    105073
           1       0.60      0.37      0.46     69233

    accuracy                           0.65    174306
   macro avg       0.64      0.60      0.60    174306
weighted avg       0.64      0.65      0.63    174306


With PCA
Best Parameters: {'criterion': 'entropy', 'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 2}
Accuracy: 0.654580659041199
Classification Report:
               precision    recall  f1-score   support

           0       0.67      0.83      0.74     83870
           1       0.61      0.38      0.47     55575

    accuracy                           0.65    139445
   macro avg       0.64      0.61      0.61    139445
weighted avg       0.64      0.65      0.63    139445

Best Parameters: {'criterion': 'entropy', 'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 10}
Accuracy: 0.6529838330292703
Classification Report:
               precision    recall  f1-score   support

           0       0.67      0.84      0.74    105073
           1       0.60      0.37      0.46     69233

    accuracy                           0.65    174306
   macro avg       0.64      0.60      0.60    174306
weighted avg       0.64      0.65      0.63    174306